### DISCUSSION TITLE
Stepper Disabling after print completes and 2 minutes after a home.

### DISCUSSION THREAD

**Discussion by I-wish-2** [mriscoc/Ender3V2S1#607] [Q&A]
Greetings All,
I think this is the correct area to ask for some help.  I have several modified E3 Pros that are using firmware Ver 2.1.3 20221222, 427-BLTUBL-MPC.  Several of these printers have had the timing belt modification by Kevin AkaSam ( YouTube).  These printers require the Z axis to remain powered up when at idle to keep the X axis Gantry from drifting down causing a nozzle impact with the bed.
I understand from Miguel that the steppers and temperature being disabled at the end of the print or after 2 minutes of inactivity after a "home all" function, is safety related.  If you all could help me with the edit Miguel suggested I would be so  grateful and maybe much smarter.
It was suggested to edit the source code file dwin.cpp, ( Marlin/src/lcd/e3v2/proui/dwin.cpp ), and then recompile and flash the mother board.  The lines of code suggested to be edited are as follows, although I'm not sure as to which ones.

1621  // Ended print job
1622  void DWIN_Print_Finished() {
1623  TERN_(DEBUG_DWIN, SERIAL_ECHOLNPGM("DWIN_Print_Finished"));
1624  TERN_(POWER_LOSS_RECOVERY, if (card.isPrinting()) recovery.cancel());
1625  HMI_flag.pause_flag = false;
1626  wait_for_heatup = false;
1627  planner.finish_and_disable();
1628  thermalManager.cooldown();
1629  if (!HMI_flag.config_flag) Goto_PrintDone(); else HMI_flag.config_flag = false;
1630  }

I reviewed a similar request in the Discussions Area for a work around regarding the hotend and bed temperature being set to 0 at the end of printing and the suggestion was to comment out line 1628 to remove that function, I do not know the results.

I guessed that line 1627, with disable in it, was the line I need to,  //,  comment out.  I was able to edit the line using Visual Studio Code and then recompiling and flashing the mother board which I did.  After flashing the firmware rebooted and came up ok but I was not able to home or mover any of the Axis's and then the home timed out and a power down was required.

I would like my End G-Code to handle the end of print requirements which is below.

; Ender 3 & 5 Custom END G-code 12/12/2022
; This code is based on the E3 V2 Klipper setup buts work on all E3 & E5 Printers

G91		                ;Relative positioning
G1 E-2 F2700	        ;Retract 2 mm
G1 E-2 Z0.2 F2400	;Retract and raise Z
G1 X5 Y5 F3000	;Wipe out
G1 Z10		        ;Raise Z more
G90		                ;Absolute positioning
G1 X10 Y225 Z75	;Park X and Z axis; Move Y to Present Print
M106 S0		        ;Turn-off fan
M104 S0		        ;Turn-off hotend
M140 S0		        ;Turn-off bed
M84 X Y E 	        ;Disable all steppers but Z to lock the gantry at 75mm

; End of custom END G-Code

I'm now at a loss as to how to proceed and any suggestions / help is more that welcome.
Doug

**Comment by mriscoc** [comment]
Use
```cpp
planner.synchronize();
```
Instead of
```cpp
planner.finish_and_disable();
```
This will fix the motor disabling after printing but not after idle.

For fix the stepper deactivation when idle change the Configuration_adv.h file:
```cpp
#define DEFAULT_STEPPER_DEACTIVE_TIME 120
#define DISABLE_INACTIVE_X true
#define DISABLE_INACTIVE_Y true
#define DISABLE_INACTIVE_Z true  // Set 'false' if the nozzle could fall onto your printed part!
```

**Reply by mriscoc** [reply]
It could be something to be configurable, for a future version.

**Reply by I-wish-2** [reply]
Greetings, thank you for direction on the compile.  I need to know how to select 4.2.7 for the compile as I'm not seeing an option for a selection of which mother board to compile for.
Doug

**Reply by mriscoc** [reply]
I think that you need this: 
![Configurator](https://user-images.githubusercontent.com/2745567/210666778-8b33249e-a81f-4e77-b77b-68284dfda0a2.png)

From: https://github.com/mriscoc/Special_Configurations

**Reply by I-wish-2** [reply]
I did miss the Professional Firmware Configurator.  It looks like a great tool.  Please understand that as a novice in this effort there are areas that I would not know were there or even how to use them.  So far I'm progressing if slowly but am having some success.  Thank you for your help.
Doug

**Comment by mriscoc** [comment]
Ok maybe we can set your configuration as a sponsored one, here it is and again, thanks for your support:

===============
Hi Doug, I'm attached a firmware with your requirements, the NCDF are for NOT_COOL_OR_DISABLE_AT_FINISH, I create that configuration for you, also the DEFAULT_STEPPER_DEACTIVE_TIME was increased to 300 s and DISABLE_INACTIVE_Z was set to false.

(file removed)

Regards.

===============
Unzip the firmware file and copy to your memory card

**Reply by mriscoc** [reply]
Please set your end g-code script to cooldown or disable the motors if you need.

**Reply by I-wish-2** [reply]
My Current End G-Code
; Ender 3 & 5 Custom END G-code 12/12/2022
; This code is based on the E3 V2 Klipper setup buts work on all E3 & E5 Printers

G91		                ;Relative positioning
G1 E-2 F2700	        ;Retract 2 mm
G1 E-2 Z0.2 F2400	;Retract and raise Z
G1 X5 Y5 F3000	;Wipe out
G1 Z10		        ;Raise Z more
G90		                ;Absolute positioning

G1 X10 Y225 Z75	;Park X and Z axis; Move Y to Present Print
M106 S0		        ;Turn-off fan
M104 S0		        ;Turn-off hotend
M140 S0		        ;Turn-off bed

M84 X Y E 	         ;Disable all steppers but Z to lock the gantry at 75mm

; End of custom END GCode

The Flash is complete and I'm coping in the configuration setting.  Temp tuning next and then some prints.  More info in a few.
Miguel Thank You,  My Boat Anchor is soon to be a printer.
Doug

**Reply by I-wish-2** [reply]
Hi Miguel,
The current results of the bin you sent.
All configuration data for prepare, control and advanced was entered ok.  The hot end and bed temp auto tune worked great.  The tramming and bed mesh at 80 deg ( PET-G filament) was near flat.  I continue to run and adjust the wheels until the tramming is 0.01 on each corner and the Mesh will follow close to those values.  The Z-Offset was the same a before.
I printed a calibration square 40 x 40 x 2 layers for the z-offset which used my End G_Code to test the Z axis.  This failed as the X Y Z & E axis all disabled and the gantry dropped.
After a home all the z axis was raised to 340 mm and after 900 seconds / 15 minutes of inactive all the stepper were still enabled.
Progress is being made Ah, so close.  I poured you a nice craft brew toasted you and called it an evening as tomorrow is another day for this adventure.  Again thank you my Friend.
My Best Regards
Doug

**Reply by mriscoc** [reply]
I forgot to say that in the Creality 422 and 427 boards, the stepper drivers for all axes share the same enable pin. I don't remember how M84 manages that, I will test that tomorrow but for now try removing the M84.

**Reply by project-nope** [reply]
> I forgot to say that in the Creality 422 and 427 boards, the stepper drivers for all axes share the same enable pin. I don't remember how M84 manages that, I will test that tomorrow but for now try removing the M84.

**This is the information I have been looking for.**  I was having the most difficult problems with M84/M18 not working correctly, and thought there was a bug in your firmware.  Thank you for explaining this!

I am changing all occurances of:
`M84 X Y E`
in my gcode to
`M84 S3600`

I do not know why, but after squaring my frame with engineering indicators and oiling my Z-axis, it is falling like a stone as soon as steppers disable.

The eccentric nuts are very tight, I can only _barely_ turn them against the rail with 2 fingers.  All bolts on the printer are torqued with 2 newton-meter torque wrench - all bolts are uniformly tight.  All extrusions are straight.

I think my Z-axis screw is too perfectly straight and well oiled, so it falls too easily.  If you have another idea I'd love to hear it!

**Comment by I-wish-2** [comment]
Here is the custom E3 Pro effort that needs the Z help.
Doug

![A 500mm Z Axis E3 Pro](https://user-images.githubusercontent.com/103605403/210686069-0cca25a5-5fb2-4e26-b59a-08c258e1bed4.jpg)
![Rear View](https://user-images.githubusercontent.com/103605403/210686082-acee3f20-99fe-454a-8c59-3880af7f324b.jpg)

**Reply by mriscoc** [reply]
> Here is the custom E3 Pro effort that needs the Z help. Doug
> 

Great work!

**Comment by I-wish-2** [comment]
I am giving that a try now, thank you again.

**Reply by I-wish-2** [reply]
Greetings,
I tested the End G-Code with M84 command in and then commented out.  The results were the same for both tests.  At the end of the print all of the Axis's were disabled.
Would testing with the M18 command have any value?

**Reply by mriscoc** [reply]
ok I just saw this in the Configuration_adv.h file:
https://github.com/mriscoc/Ender3V2S1/blob/72b609366965ba83826b6124878193de20c3d664/Marlin/Configuration_adv.h#L1542-L1543
So, Marlin core also disables the stepper automatically at the end of SD printing, I must comment out these lines and compile again.

**Reply by I-wish-2** [reply]
That is great news, maybe the resolution is in sight.  I have tested the inactivity time out again and it appears that the steppers are still active after 1 hour, my longest test time.  This is not an issue for me because I turn off the Hot End and Bed when I'm at idle for any amount of time.  The 600 seconds would be a good safety for sure.

**Reply by mriscoc** [reply]
> The 600 seconds would be a good safety for sure.

Seems that set `DEFAULT_STEPPER_DEACTIVE_TIME` to 600 will not work due to the shared enable pin, if we set to false `DISABLE_INACTIVE_Z` the other axes will keep enabled also.

https://github.com/mriscoc/Ender3V2S1/blob/72b609366965ba83826b6124878193de20c3d664/Marlin/Configuration_adv.h#L1123-L1126

**Reply by mriscoc** [reply]
Maybe it would be better to leave `DISABLE_INACTIVE_Z` to true and set a longer time for DEFAULT_STEPPER_DEACTIVE_TIME, what do you think?

**Reply by I-wish-2** [reply]
The idle test ran for about 12 minutes and the hot end, bed and all the steppers remained active. 
" Maybe it would be better to leave DISABLE_INACTIVE_Z to true and set a longer time for DEFAULT_STEPPER_DEACTIVE_TIME "  This now seems to be the best way to ensure the safety of all makers.

**Reply by I-wish-2** [reply]
" Seems that set DEFAULT_STEPPER_DEACTIVE_TIME to 600 will not work due to the shared enable pin, if we set to false DISABLE_INACTIVE_Z the other axes will keep enabled also. "
Do I understand that if the DISABLE_INACTIVE_Z is set to TRUE then the timer will function at 600 seconds correctly?  If so this seems to be the best way to set up the INACTIVE state.

**Reply by I-wish-2** [reply]
Hi, sorry for the 600 sec reference.  I don't know where I picked that up from.  The 300 sec that you mentioned is better.  Although the current bin version does not deactivate steppers or heater elements.

**Reply by mriscoc** [reply]
OK, I sent you a new version, configuration applied:

```json
{
"Configuration.h" : [
  {
    "op": "InsertAfter",
    "searchfor": "#define DWIN_LCD_PROUI",
    "newline": "#define NOT_COOL_OR_DISABLE_AT_FINISH\n"
  }
],
"Configuration_adv.h" : [
  {
    "op": "CustomVal",
    "searchfor": "DEFAULT_STEPPER_DEACTIVE_TIME",
    "value": "300"
  },
  {
    "op": "Disable",
    "searchfor": "SD_FINISHED_STEPPERRELEASE"
  }
],
"Version.h" : [
]   
}
```

**Reply by I-wish-2** [reply]
Greetings,  Just back in and thank you for the new version.  I will load it and test and get the results back to you when completed.
Miguel thank you for taking the time to help me, it means a lot as I know your time is limited.
Regards
Doug

**Reply by I-wish-2** [reply]
And for the new configuration that are applied.
D

**Reply by I-wish-2** [reply]
Good Morning Miguel,
Here are the test results for the new version 20230105.                    A general Note to all:

I strongly recommend that you " Auto Home " your printer prior to doing any type of movement calibrations or temperature tuning to ensure correct axis positioning at the beginning of your calibration / tuning.

The Inactive State setup test 1.  Auto Home All   >  Move Z to 250mm  >   Using " Control " Temperature - Hot End Temp = 165.  Bed Temp = 80.   Start the Inactive test 1.

TEST 1 Results:  At around 297 seconds all the steppers disabled.  The hot end and bed remained at their respective setpoints.

The Inactive State setup test 2.  Auto Home All   >  Move Z to 250mm  >   Using " Prepare " Preheat CUSTOM- Hot End Temp = 165.  Bed Temp = 80.   Start the Inactive test 2.

TEST 2 Results:  At about 297 seconds all the steppers disabled.  The hot end and bed remained at their respective setpoints.

End of Inactive configuration testing.

Start of Print configuration Tests:

TEST 1,  My End G-Code with the Fan, Hotend, Bed and all stepper commands COMMENTED OUT.

G91		;Relative positioning
G1 E-2 F2700	;Retract 2 mm
G1 E-2 Z0.2 F2400	;Retract and raise Z
G1 X5 Y5 F3000	;Wipe out
G1 Z10		;Raise Z more
G90		;Absolute positioning

G1 X10 Y225 Z75	;Park X and Z axis; Move Y to Present Print

;M106 S0		;Turn-off fan
;M104 S0		;Turn-off hotend
;M140 S0		;Turn-off bed
;M84 X Y E 	;Disable all steppers but Z to lock the gantry at 75mm
; End of custom END G-Code

Print test 1 results.  No steppers were disabled, the Hot End 240 degrees and the Bed 80 degrees remained at their setpoints.

TEST 2,  My End G-Code with the Fan, Hotend, Bed and all Stepper commands are ACTIVE.

G91		;Relative positioning
G1 E-2 F2700	;Retract 2 mm
G1 E-2 Z0.2 F2400	;Retract and raise Z
G1 X5 Y5 F3000	;Wipe out
G1 Z10		;Raise Z more
G90		;Absolute positioning

G1 X10 Y225 Z75	;Park X and Z axis; Move Y to Present Print

M106 S0		;Turn-off fan
M104 S0		;Turn-off hotend
M140 S0		;Turn-off bed
M84 X Y E 	;Disable all steppers but Z to lock the gantry at 75mm
; End of custom END G-Code

Print test 2 results.  No steppers were disabled, the Hot End 240 degrees and the Bed 80 degrees remained at their setpoints.

Comments:

For the Inactive 300 second configuration I feel it would be best if the Hot End and the Bed setpoints were set to "0".  This would insure that if the any axis drifted and made bed or other types of contact it would be with the elements cooling off.

For the End of Print configuration these results are what I needed for my printer.  I must say though that if the Hot End and the Bed setpoints were set to "0" that would be good safety wise.

Miguel thank you for your time and effort with this configuration and please let me know what you think of the results or if there is any other testing that you would like me to do.
Doug

**Reply by mriscoc** [reply]
The Timeout for the heaters is defined as 10 minutes after the last extruder movement.

https://github.com/mriscoc/Ender3V2S1/blob/c2086647a8ba318829073bce60e00afaa4955d68/Marlin/Configuration_adv.h#L513-L523

**Reply by I-wish-2** [reply]
HI Miguel, 
Sorry for the delay in replying.  Sometimes Life is what happens while your busy planning other things.
Thank you for the info regarding the heater time out triggered by the last Extruder movement.  That is a great safety.

Based on the results my testing with the Experimental-Ender3V2-BLTUBL-MPC-NCDF 20230105.  I wanted to control the heaters during my End G-Code cycle.  After some trial and effort I came up with the following End G-Code. it will cool down the Hot End and Bed in two steps and holds the steppers enabled for over 10 minutes after the End G-Code completes.

At the end of the print the Hot End is at 240 Degrees, the Bed is at 80 Degrees and all steppers are enabled.  The End G-Code then starts.     The End G-Code is as follows:

; Ender 3 & 5 Custom END G-code 01/07/2023

G91;                                  Relative positioning
G1 E-3 F2700;                   Retract 2 mm
G1 E-2 Z0.2 F2400;           Retract and raise Z
G1 X5 Y5 F300;                 Wipe out
G1 Z10;                             Raise Z more
G90;                                  Absolute positioning
G1 X10 Y195 Z75 F2000;  Park X and Z axis; Move Y to Present Print.
M106 S255;                      Turn fan full on to help the cool down the hot end.
M109 R165;                      Wait for Hot End to cool to 165 Degrees.
M107;                               Turn-off fan
M190 R70;                        Wait for the Bed to cool to 70 Degrees.
M104 S0;                          Turn-off hotend
M140 S0;                          Turn-off bed
G4 240;                            Dwell delay of 4 minutes for the print cooling and removal.                     

; End of custom END G-Code

These 14 lines of code had this result.

Lines 1 thru 6      move the Hot End and all Steppers from the print.
Line 7                  parks the Hot End and Y presents the print to the front of the printer.
Line 8                  turns on the cooling fan 100%.
Line 9                  starts ramping down the Hot End to 165 degrees.               about 2 minutes to line 10
Line 10                turns off the cooling fan, Hot End is now 165.
Line 11                starts ramping down the Bed temperature to 70 degrees.   about 2:30 minutes to line 
Line 12                sets the Hot End to " 0 " degrees.
Line 13                sets the Bed to " 0 " degrees.
Line 14                starts a dwell ( delay ) of 4 minutes.

In about 14 minutes all the steppers are disabled.

Miguel with your 20230105 version and the above End G-Code, I feel you have gotten me to a functional and safe operational point with my E3 Pro Timing Belt printer.  When I thought about the Inactivity setting of 300 seconds I agree that If your working with or testing the printer you will never be at the 5 minute mark without making a change of some kind.

I think that we can put this one to rest, unless you have anything else to test or review.
Again My Best Regards
Doug

**Reply by mriscoc** [reply]
Why don't you turn off the hot-end and bed just after raise Z?

**Reply by I-wish-2** [reply]
Great thought let me give it a try, Thank you.
D

**Reply by I-wish-2** [reply]
Sometimes I over think it.  This is much better Thank You!

; Ender 3 Pro Z Axis Timing Belt Drive
; Custom END G-code 01/07/2023

G91			        ;Relative positioning
G1 E-3 F2700		;Retract 3 mm
G1 Z0.4 E-2 F2400   ;Raise Z and Retract again
G1 X5 Y5 F600		;Wipe out
G1 Z10			;Raise Z again
M104 S0			;Turn-off hotend
M140 S0			;Turn-off bed
M107			;Turn-off fan
G90			        ;Absolute positioning
G1 X10 Y195 Z75 F2400	;Park X and Z axis; Move Y to Present Print.

; End of custom END G-Code

Doug

**Reply by I-wish-2** [reply]
Hi Miguel,
With more that a few prints run now the printer is preforming better than ever as I have not seen better print quality up to this point.
Is it possible to receive a copy of the source code that you used to compile the version of "Experimental-Ender3v2-427-BLTUBL-MPC-NCDF-20230105.bin " for me.  I would like to keep a copy in my Firmware Backup folder if it is needed.  Once again thank you for all your help.
Doug

**Reply by mriscoc** [reply]
Of course, and better if you can always have the later firmware version just by using the **Configurator.pyw**, copy the text bellow and save in the **_features** folder with the filename: `NCDF.json`, the Configurator will be see it as a new feature and you will can select it.
```json
{
"Configuration.h" : [
  {
    "op": "InsertAfter",
    "searchfor": "#define DWIN_LCD_PROUI",
    "newline": "#define NOT_COOL_OR_DISABLE_AT_FINISH\n"
  }
],
"Configuration_adv.h" : [
  {
    "op": "CustomVal",
    "searchfor": "DEFAULT_STEPPER_DEACTIVE_TIME",
    "value": "300"
  },
  {
    "op": "Disable",
    "searchfor": "SD_FINISHED_STEPPERRELEASE"
  }
],
"Version.h" : [
]   
}
```

**Comment by I-wish-2** [comment]
Thank you Miguel,

I have tried to figure out how to use the Configurator.pyw file and Pylon 3.1 that I have installed with no luck so far.  I have started a discussion on your site and enlisted any and all help.  I did review any thread that had Configurator in it but there was not much help there. I’m looking forward to any answers as there are others who have asked for similar information.  

Best Regards

Doug

p.s. my prints are near perfect as they look almost like injection molded items. : )
